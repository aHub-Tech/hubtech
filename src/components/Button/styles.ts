import styled, { css } from 'styled-components'
import T from 'theme'
import { BtnProps } from '.'

const SizesBtn = {
  xxSmall: css`
    padding: 0.5rem 0.6rem;
    font-size: ${T.typography.sizes.p4};
  `,
  xSmall: css`
    padding: 0 0.6rem;
    font-size: ${T.typography.sizes.p4};
  `,
  small: css`
    font-size: ${T.typography.sizes.p4};
    padding: 0.8rem;
  `,
  medium: css`
    font-size: ${T.typography.sizes.p4};
    padding: 1rem 2.5rem;
  `,
  large: css`
    font-size: ${T.typography.sizes.p3};
    padding: 1.2rem 4rem;
  `,
  xLarge: css`
    font-size: ${T.typography.sizes.p1};
    padding: 1.5rem 4rem;
  `
}

export const Container = styled.button<BtnProps>`
  ${({ fullWidth, fullWidthMobile, color, size, borderRadiusFull }) => css`
    cursor: pointer;
    border-radius: ${borderRadiusFull ? '10rem' : '0.5rem'};
    font-weight: 400;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    flex-wrap: nowrap;
    border: 0.1rem solid transparent;
    text-shadow: 0 0 0.3rem #00000099;
    background-position-x: 0;
    background-size: 300% !important;
    background-repeat: no-repeat;
    transition: background-position-x 150ms ease-in-out;
    width: ${fullWidthMobile ? '100%' : 'fit-content'};
    background-color: ${!!color && T.colors[color!]};
    ${!!size && SizesBtn[size]}

    ${T.breakPoint(T.devices.xs)} {
      width: ${fullWidth ? '100%' : 'fit-content'};
    }

    &:hover {
      opacity: 0.9;
    }
    &:focus {
      box-shadow: 0 0 0 0.1rem #fff;
    }

    &:active {
      box-shadow: none;
    }

    &:disabled,
    &:disabled:hover {
      opacity: 0.5;
      cursor: not-allowed;
      background-position-x: 0;
    }

    span {
      color: ${T.colors.white};
    }

    svg + span {
      margin-left: 1rem;
    }

    svg {
      width: 2rem;
      min-width: 2rem;
      height: 2rem;
      min-height: 2rem;
      color: ${T.colors.white};
    }
  `}
`
