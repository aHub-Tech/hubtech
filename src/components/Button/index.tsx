import { ButtonHTMLAttributes, forwardRef, ReactNode } from 'react'
import * as S from './styles'

export type SizeProps =
  | 'xxSmall'
  | 'xSmall'
  | 'small'
  | 'medium'
  | 'large'
  | 'xLarge'

export type ColorProps =
  | 'primary'
  | 'facebook'
  | 'discord'
  | 'instagram'
  | 'linkedin'
  | 'twitch'
  | 'github'
  | 'youtube'
  | 'link'

export type BtnProps = {
  icon?: ReactNode
  text?: string
  size?: SizeProps
  color?: ColorProps
  fullWidth?: boolean
  fullWidthMobile?: boolean
  borderRadiusFull?: boolean
  className?: string
  isLink?: boolean
  href?: string
} & ButtonHTMLAttributes<HTMLButtonElement>

const Button = forwardRef<HTMLButtonElement, BtnProps>(
  (
    {
      icon,
      text,
      size = 'xSmall',
      color = 'primary',
      fullWidth,
      fullWidthMobile,
      isLink = false,
      href = '',
      ...props
    },
    ref
  ) => (
    <S.Container
      as={isLink ? 'a' : 'button'}
      color={color}
      size={size}
      fullWidth={fullWidth}
      fullWidthMobile={!!fullWidth || fullWidthMobile}
      {...props}
      {...(isLink && { href, target: '_blank' })}
      ref={ref}
    >
      {icon}
      {!!text && <span>{text}</span>}
    </S.Container>
  )
)

export default Button
