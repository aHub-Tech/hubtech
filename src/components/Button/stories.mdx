import { Canvas, Meta, Story } from '@storybook/addon-docs'
import Button from '.'
import * as stories from './stories'

<Meta
  title="Design System/Button"
  component={Button}
  args={{
    text: 'Clique aqui',
    title: 'Instructions',
    size: 'xLarge',
    color: 'blue',
    disabled: false,
    fullWidth: false,
    fullWidthMobile: false
  }}
  argTypes={{
    size: {
      options: ['xxSmall', 'xSmall', 'small', 'medium', 'large', 'xLarge'],
      control: { type: 'inline-radio' }
    },
    color: {
      options: ['red', 'green', 'blue'],
      control: { type: 'inline-radio' }
    },
    onClick: { action: 'onClick' }
  }}
  parameters={{ controls: { exclude: ['icon'], sort: 'requiredFirst' } }}
/>

# Button

This is a general purpose button component.

### Default

The button receives props for styling, icons, as well as a click event,
and makes its reference available through the forwardRef() hook.

<Canvas>
  <Story story={stories.Default} />
</Canvas>
