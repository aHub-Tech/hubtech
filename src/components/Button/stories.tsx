import { Story } from '@storybook/react'
import { Eye } from '@styled-icons/ionicons-outline'
import styled from 'styled-components'
import Button, { BtnProps } from '.'

const WrapperBtn = styled.div`
  display: flex;
  gap: 2rem;
  flex-wrap: wrap;
  justify-content: center;
`

export const Template: Story<BtnProps> = args => (
  <WrapperBtn>
    <Button {...args} />
    <Button {...args} icon={<Eye />} />
    <Button {...args} text="" icon={<Eye />} />
    <Button {...args} text="" icon={<Eye />} />
    <Button {...args} text="" icon={<Eye />} />
    <Button {...args} text="" icon={<Eye />} />
  </WrapperBtn>
)

export const Default = Template.bind({})
