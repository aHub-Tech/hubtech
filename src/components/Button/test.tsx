import { Eye } from '@styled-icons/ionicons-outline'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import T from 'theme'
import { renderWithTheme } from 'utils/tests/helpers'
import Button from '.'

describe('<Button />', () => {
  it('should render the xSmall size by default', () => {
    const { container } = renderWithTheme(<Button text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      padding: '0px 0.6rem',
      fontSize: 'xSmall'
    })

    expect(container).toMatchSnapshot()
  })

  it('should render the xxSmall size', () => {
    renderWithTheme(<Button size="xxSmall" text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      padding: '0.5rem 0.6rem',
      fontSize: T.typography.sizes.p4
    })
  })

  it('should render the small size', () => {
    renderWithTheme(<Button size="small" text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      fontSize: T.typography.sizes.p4,
      padding: '0.8rem'
    })
  })

  it('should render the medium size', () => {
    renderWithTheme(<Button size="medium" text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      fontSize: T.typography.sizes.p4,
      padding: '1rem 2.5rem'
    })
  })

  it('should render the large size', () => {
    renderWithTheme(<Button size="large" text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      fontSize: T.typography.sizes.p3,
      padding: '1.2rem 2rem'
    })
  })

  it('should render the xLarge size', () => {
    renderWithTheme(<Button size="xLarge" text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      fontSize: T.typography.sizes.p1,
      padding: '1.5rem 3rem'
    })
  })

  it('should render a fullWidth version', () => {
    renderWithTheme(<Button fullWidth text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<Eye data-testid="icon" />} text="Mais Informações" />
    )

    const btn = screen.getByText(/Mais Informações/i)
    const icon = screen.getByTestId('icon')
    expect(btn).toBeInTheDocument()
    expect(icon).toBeInTheDocument()
  })

  it('should render a disabled Button', () => {
    renderWithTheme(<Button disabled text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      cursor: 'not-allowed',
      opacity: 0.5,
      backgroundPositionX: 0
    })
  })

  it('should call a function once on click', () => {
    const onClick = jest.fn()

    renderWithTheme(<Button onClick={onClick} text="Mais Informações" />)

    const btn = screen.getByRole('button', { name: /Mais Informações/i })

    userEvent.click(btn)

    expect(onClick).toHaveBeenCalled()
    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
