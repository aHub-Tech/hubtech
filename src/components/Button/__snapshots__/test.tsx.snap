// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the xSmall size by default 1`] = `
.c0 {
  cursor: pointer;
  border-radius: 0.5rem;
  font-weight: 400;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  border: 0.1rem solid transparent;
  text-shadow: 0 0 0.3rem #00000099;
  background-position-x: 0;
  background-size: 300% !important;
  background-repeat: no-repeat;
  -webkit-transition: background-position-x 150ms ease-in-out;
  transition: background-position-x 150ms ease-in-out;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  background-color: #F83B91;
  padding: 0 0.6rem;
  font-size: 1.2rem;
}

.c0:hover {
  opacity: 0.9;
}

.c0:focus {
  box-shadow: 0 0 0 0.1rem #fff;
}

.c0:active {
  box-shadow: none;
}

.c0:disabled,
.c0:disabled:hover {
  opacity: 0.5;
  cursor: not-allowed;
  background-position-x: 0;
}

.c0 span {
  color: #fff;
}

.c0 svg + span {
  margin-left: 1rem;
}

.c0 svg {
  width: 2rem;
  min-width: 2rem;
  height: 2rem;
  min-height: 2rem;
  color: #fff;
}

@media screen and (min-width:500px) {
  .c0 {
    width: -webkit-fit-content;
    width: -moz-fit-content;
    width: fit-content;
  }
}

<div>
  <button
    class="c0"
    color="primary"
  >
    <span>
      Mais Informações
    </span>
  </button>
</div>
`;
